src/qibolab/identifier.py

```diff
@@ -0,0 +1,63 @@
+from enum import Enum
+from typing import Annotated, Optional, Union
+
+from pydantic import Field, TypeAdapter, model_serializer, model_validator
+
+from .serialize import Model
+
+QubitId = Annotated[Union[int, str], Field(union_mode="left_to_right")]
+"""Type for qubit names."""
+
+
+# TODO: replace with StrEnum, once py3.10 will be abandoned
+# at which point, it will also be possible to replace values with auto()
+class ChannelType(str, Enum):
+    """Names of channels that belong to a qubit.
+
+    Not all channels are required to operate a qubit.
+    """
+
+    PROBE = "probe"
+    ACQUISITION = "acquisition"
+    DRIVE = "drive"
+    DRIVE12 = "drive12"
+    DRIVE_CROSS = "drive_cross"
+    FLUX = "flux"
+
+    def __str__(self) -> str:
+        return str(self.value)
+
+
+class ChannelId(Model):
+    """Unique identifier for a channel."""
+
+    qubit: QubitId
+    channel_type: ChannelType
+    cross: Optional[str]
+
+    @model_validator(mode="before")
+    @classmethod
+    def _load(cls, ch: str) -> dict:
+        elements = ch.split("/")
+        # TODO: replace with pattern matching, once py3.9 will be abandoned
+        if len(elements) > 3:
+            raise ValueError()
+        q = TypeAdapter(QubitId).validate_python(elements[0])
+        ct = ChannelType(elements[1])
+        assert len(elements) == 2 or ct is ChannelType.DRIVE_CROSS
+        dc = elements[2] if len(elements) == 3 else None
+        return dict(qubit=q, channel_type=ct, cross=dc)
```

Comment:
> The main thing I don't fully like is that we are sort of programming via strings and this results to the usual ugly parser (see `PulseShape` and `eval` in 0.1, just different kind here).

They are connected because both related to (de)serialization, but `PulseShape` was a much more structured object (a full pulse), while this is pretty simpler (the complicate part is that they are 2 or 3 sections, and that's variable - without `drive_cross` it would be much easier).

However, I also don't like it too much, because I'm "manually" parsing, while it could be done automatically by Pydantic (which, unless there are Python functions in the middle, is generating a JSON schema, and then processing the JSON in Rust).

To be fair, `drive_cross` is also a bit silly, because is not a channel associated to a single qubit, but more to a pair...
(conceptually, then it will play on a physical line, but it's relevant up to a point whether it is connected to one qubit or the other, it's just addressing the pair)

I'd like much more to have `chtype/something`, where something may contain the qubit, but also a qubit pair and what not (qudits, if we go beyond qutrits, and whatever else).
As said above, the association `channel <-> qubit` could be kept separate from the channel retrieval.
Moreover, `chtype` doesn't even to be that granular, but it's just convenient to have the same information encoded by the channel classes locally in the sequence, i.e. separate DC, modulated, and acquisition.

Reply:
> As said above, the association `channel <-> qubit` could be kept separate from the channel retrieval.

I also don't understand this. To me it seems that now we cannot even initialize a `Channel` without refering to a specific `QubitId`.

Reply:
In #1001 I'll try to go back to plain strings, without any special meaning.